src/bin/ipa_bench/gen.rs

```diff
@@ -0,0 +1,428 @@
+use super::sample::Sample;
+use byteorder::ByteOrder;
+use chrono::Duration;
+use log::{debug, info, trace};
+use rand::SeedableRng;
+use rand::{CryptoRng, Rng, RngCore};
+use rand_chacha::ChaCha20Rng;
+use rand_distr::{num_traits::ToPrimitive, Bernoulli, Distribution};
+use raw_ipa::helpers::models::{
+    Event as EEvent, SecretShare, SourceEvent as ESourceEvent, TriggerEvent as ETriggerEvent,
+};
+use serde::{Deserialize, Serialize};
+use std::io;
+
+const DAYS_IN_EPOCH: i64 = 7;
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct Event {
+    matchkeys: Vec<u64>,
+    epoch: u8,
+    timestamp: u64,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct SourceEvent {
+    event: Event,
+    breakdown_key: String,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct TriggerEvent {
+    event: Event,
+    value: u32,
+    zkp: String,
+}
+
+enum EventType {
+    S(SourceEvent),
+    T(TriggerEvent),
+    ES(ESourceEvent),
+    ET(ETriggerEvent),
+}
+
+// TODO: Currently, users are mutually exclusive in each ad loop (i.e. User A in ad X will never appear in other ads).
+// We need to generate events from same users across ads (but how often should a user appear in different ads?)
+// "Ads" doesn't mean FB's L3 ads. It could be ads from different businesses.
+
+pub fn generate_events(
+    total_count: u32,
+    epoch: u8,
+    secret_share: bool,
+    seed: &Option<u64>,
+    out: &mut Box<dyn io::Write>,
+) -> (u32, u32) {
+    let mut rng = match seed {
```

Comment:
we should probably inject rng here

Reply:
I'm using the seed to generate two RNGs. I thought it also helps isolation of logics from 'cmd.rs'. Should I inject two RNGs instead?

Reply:
I am curious why it is required to use two RNGs here? if they're seeded with the same initial value, both produce the same output. 

Injecting RNGs may make it easier to test and control the output of RNG as well as pick a different RNG implementation based on caller's need